pkg/updater/jobparser.go

```diff
@@ -0,0 +1,331 @@
+func setDefaultAndValidate(job *paddlev1.TrainingJob) error {
+	// Fill in default values
+	// FIXME: Need to test. What is the value if specified "omitempty"
+	if job.Spec.Port == 0 {
+		job.Spec.Port = 7164
+	}
+	if job.Spec.PortsNum == 0 {
+		job.Spec.PortsNum = 1
+	}
+	if job.Spec.PortsNumForSparse == 0 {
+		job.Spec.PortsNumForSparse = 1
+	}
+	if job.Spec.Image == "" {
+		job.Spec.Image = "paddlepaddle/paddlecloud-job"
+	}
+	if job.Spec.Passes == 0 {
+		job.Spec.Passes = 1
+	}
+
+	if !job.Spec.FaultTolerant && job.Elastic() {
+		return errors.New("max-instances should equal to min-instances when fault_tolerant is disabled")
+	}
+	// TODO: add validations.(helin)
+	return nil
+}
+
+// NewTrainingJob generates a whole structure of TrainingJob
+func (p *DefaultJobParser) NewTrainingJob(job *paddlev1.TrainingJob) (*paddlev1.TrainingJob, error) {
+	if err := setDefaultAndValidate(job); err != nil {
+		return nil, err
+	}
+
+	useHostNetwork := job.Spec.HostNetwork
+	if job.Spec.FaultTolerant {
+		job.Spec.Master.ReplicaSpec = parseToMaster(job)
+		if useHostNetwork {
+			job.Spec.Master.ReplicaSpec.Spec.Template.Spec.HostNetwork = true
+		}
+	}
+	job.Spec.Pserver.ReplicaSpec = parseToPserver(job)
+	job.Spec.Trainer.ReplicaSpec = parseToTrainer(job)
+	if useHostNetwork {
+		job.Spec.Pserver.ReplicaSpec.Spec.Template.Spec.HostNetwork = true
+		job.Spec.Trainer.ReplicaSpec.Spec.Template.Spec.HostNetwork = true
+	}
+	return job, nil
+}
+
+// parseToPserver generate a pserver replicaset resource according to "TrainingJob" resource specs.
+func parseToPserver(job *paddlev1.TrainingJob) *v1beta1.ReplicaSet {
+	replicas := int32(job.Spec.Pserver.MinInstance)
+	command := make([]string, 2, 2)
+	// FIXME: refine these part.(typhoonzero)
+	if job.Spec.FaultTolerant {
+		command = []string{"paddle_k8s", "start_pserver"}
+	} else {
+		command = []string{"paddle_k8s", "start_new_pserver"}
+	}
+
+	return &v1beta1.ReplicaSet{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "extensions/v1beta1",
+			APIVersion: "ReplicaSet",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      job.ObjectMeta.Name + "-pserver",
+			Namespace: job.ObjectMeta.Namespace,
+		},
+		Spec: v1beta1.ReplicaSetSpec{
+			Replicas: &replicas,
+			Template: corev1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: map[string]string{"paddle-job-pserver": job.ObjectMeta.Name},
+				},
+				Spec: corev1.PodSpec{
+					Volumes: job.Spec.Volumes,
+					Containers: []corev1.Container{
+						corev1.Container{
+							Name:      "pserver",
+							Image:     job.Spec.Image,
+							Ports:     podPorts(job),
+							Env:       podEnv(job),
+							Command:   command,
+							Resources: job.Spec.Pserver.Resources,
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+// parseToTrainer parse TrainingJob to a kubernetes job resource.
+func parseToTrainer(job *paddlev1.TrainingJob) *batchv1.Job {
+	replicas := int32(job.Spec.Trainer.MinInstance)
+	command := make([]string, 2)
```

Comment:
`make([]string, 2)` creates a slice which is never used. `var command []string` is suffice.

Reply:
done.